File: js/plugins/otel/tests/gen-ai-instrumentation_test.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { SpanKind, type Span } from '@opentelemetry/api';
import {
  InMemorySpanExporter,
  type ReadableSpan,
} from '@opentelemetry/sdk-trace-base';
import * as assert from 'assert';
import type {
  GenkitSpanContext,
  InstrumentationSpanInfo,
} from 'genkit/tracing';
import { beforeEach, describe, it } from 'node:test';
import { GenAiInstrumentation } from '../src/gen-ai-instrumentation.js';
import { getTracer } from './utils.js';

const exporter = new InMemorySpanExporter();
const tracer = getTracer(exporter);

function info(overrides: Partial<InstrumentationSpanInfo> = {}) {
  return {
    metadata: { name: 'test', ...(overrides.metadata ?? {}) },
    labels: overrides.labels ?? {},
  } as InstrumentationSpanInfo;
}

async function runAndGetSpan(
  inst: GenAiInstrumentation,
  spanInfo: InstrumentationSpanInfo,
  result: unknown
): Promise<ReadableSpan> {
  await inst.runInNewSpan(spanInfo, async (_span: Span, _ctx) => result);
  const spans = exporter.getFinishedSpans();
  return spans[spans.length - 1];
}

describe('GenAiInstrumentation', () => {
  beforeEach(() => exporter.reset());

  it('creates a chat span for model actions', async () => {
    const inst = new GenAiInstrumentation({ tracer });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: {
          name: 'googleai/gemini-flash-latest',
          input: {
            messages: [{ role: 'user', content: [{ text: 'hi' }] }],
            config: { temperature: 0.5, maxOutputTokens: 100 },
          },
        },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      { finishReason: 'stop', usage: { inputTokens: 3, outputTokens: 7 } }
    );

    assert.strictEqual(span.name, 'chat gemini-flash-latest');
    assert.strictEqual(span.kind, SpanKind.CLIENT);
    assert.strictEqual(span.attributes['gen_ai.operation.name'], 'chat');
    assert.strictEqual(
      span.attributes['gen_ai.request.model'],
      'gemini-flash-latest'
    );
    assert.strictEqual(span.attributes['gen_ai.provider.name'], 'gcp.gemini');
    assert.strictEqual(span.attributes['gen_ai.request.temperature'], 0.5);
    assert.strictEqual(span.attributes['gen_ai.request.max_tokens'], 100);
    assert.deepStrictEqual(span.attributes['gen_ai.response.finish_reasons'], [
      'stop',
    ]);
    assert.strictEqual(span.attributes['gen_ai.usage.input_tokens'], 3);
    assert.strictEqual(span.attributes['gen_ai.usage.output_tokens'], 7);
  });

  it('reports tool_calls finish reason when a tool request is present', async () => {
    const inst = new GenAiInstrumentation({ tracer });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: { name: 'googleai/gemini-flash-latest' },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      {
        finishReason: 'stop',
        message: {
          role: 'model',
          content: [{ toolRequest: { name: 'getWeather', input: {} } }],
        },
      }
    );
    assert.deepStrictEqual(span.attributes['gen_ai.response.finish_reasons'], [
      'tool_calls',
    ]);
  });

  it('emits execute_tool spans only when enabled', async () => {
    const toolInfo = info({
      metadata: { name: 'getWeather' },
      labels: { 'genkit:metadata:subtype': 'tool' },
    });

    const off = new GenAiInstrumentation({ tracer });
    const genericSpan = await runAndGetSpan(off, toolInfo, 'ok');
    assert.strictEqual(genericSpan.name, 'getWeather');
    assert.strictEqual(genericSpan.attributes['genkit.action.type'], 'tool');

    exporter.reset();
    const on = new GenAiInstrumentation({ tracer, emitToolSpans: true });
    const toolSpan = await runAndGetSpan(on, toolInfo, 'ok');
    assert.strictEqual(toolSpan.name, 'execute_tool getWeather');
    assert.strictEqual(
      toolSpan.attributes['gen_ai.operation.name'],
      'execute_tool'
    );
    assert.strictEqual(toolSpan.attributes['gen_ai.tool.name'], 'getWeather');
  });

  it('tags generic spans with the genkit action type', async () => {
    const inst = new GenAiInstrumentation({ tracer });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: { name: 'myFlow' },
        labels: { 'genkit:metadata:subtype': 'flow' },
      }),
      'ok'
    );
    assert.strictEqual(span.name, 'myFlow');
    assert.strictEqual(span.kind, SpanKind.INTERNAL);
    assert.strictEqual(span.attributes['genkit.action.type'], 'flow');
  });

  it('falls back to genkit:type when there is no subtype', async () => {
    // Directly-wrapped spans (generate, dotprompt, ...) carry only genkit:type.
    const inst = new GenAiInstrumentation({ tracer });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: { name: 'generate' },
        labels: { 'genkit:type': 'util' },
      }),
      'ok'
    );
    assert.strictEqual(span.name, 'generate');
    assert.strictEqual(span.attributes['genkit.action.type'], 'util');
  });

  it('captures content on the span in SPAN_ONLY mode', async () => {
    const inst = new GenAiInstrumentation({
      tracer,
      contentCapturingMode: 'SPAN_ONLY',
    });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: {
          name: 'googleai/gemini-flash-latest',
          input: {
            messages: [
              { role: 'system', content: [{ text: 'be nice' }] },
              { role: 'user', content: [{ text: 'hi' }] },
            ],
          },
        },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      {
        finishReason: 'stop',
        message: { role: 'model', content: [{ text: 'hello' }] },
      }
    );
    const inputMessages = JSON.parse(
      span.attributes['gen_ai.input.messages'] as string
    );
    assert.strictEqual(inputMessages[0].role, 'user');
    const systemInstructions = JSON.parse(
      span.attributes['gen_ai.system_instructions'] as string
    );
    assert.strictEqual(systemInstructions[0].content, 'be nice');
    const outputMessages = JSON.parse(
      span.attributes['gen_ai.output.messages'] as string
    );
    assert.strictEqual(outputMessages[0].role, 'assistant');
    assert.strictEqual(outputMessages[0].finish_reason, 'stop');
  });

  it('reads message and finish reason from legacy candidates[0]', async () => {
    const inst = new GenAiInstrumentation({
      tracer,
      contentCapturingMode: 'SPAN_ONLY',
    });
    // Legacy plugin shape: no top-level message/finishReason, only candidates.
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: {
          name: 'googleai/gemini-flash-latest',
          input: { messages: [{ role: 'user', content: [{ text: 'hi' }] }] },
        },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      {
        usage: { inputTokens: 3, outputTokens: 7 },
        candidates: [
          {
            index: 0,
            finishReason: 'stop',
            message: { role: 'model', content: [{ text: 'hello there' }] },
          },
        ],
      }
    );
    assert.deepStrictEqual(span.attributes['gen_ai.response.finish_reasons'], [
      'stop',
    ]);
    const outputMessages = JSON.parse(
      span.attributes['gen_ai.output.messages'] as string
    );
    assert.strictEqual(outputMessages[0].role, 'assistant');
    assert.strictEqual(outputMessages[0].parts[0].content, 'hello there');
    assert.strictEqual(outputMessages[0].finish_reason, 'stop');
  });

  it('does not capture content by default (NO_CONTENT)', async () => {
    const inst = new GenAiInstrumentation({ tracer });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: {
          name: 'googleai/gemini-flash-latest',
          input: { messages: [{ role: 'user', content: [{ text: 'hi' }] }] },
        },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      { finishReason: 'stop' }
    );
    assert.strictEqual(span.attributes['gen_ai.input.messages'], undefined);
  });

  it('does not set span content attributes in EVENT_ONLY mode', async () => {
    const inst = new GenAiInstrumentation({
      tracer,
      contentCapturingMode: 'EVENT_ONLY',
    });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: {
          name: 'googleai/gemini-flash-latest',
          input: { messages: [{ role: 'user', content: [{ text: 'hi' }] }] },
        },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      {
        finishReason: 'stop',
        message: { role: 'model', content: [{ text: 'hello' }] },
      }
    );
    // Content goes to the log event, not span attributes.
    assert.strictEqual(span.attributes['gen_ai.input.messages'], undefined);
    assert.strictEqual(span.attributes['gen_ai.output.messages'], undefined);
  });

  it('sets span content attributes in SPAN_AND_EVENT mode', async () => {
    const inst = new GenAiInstrumentation({
      tracer,
      contentCapturingMode: 'SPAN_AND_EVENT',
    });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: {
          name: 'googleai/gemini-flash-latest',
          input: { messages: [{ role: 'user', content: [{ text: 'hi' }] }] },
        },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      {
        finishReason: 'stop',
        message: { role: 'model', content: [{ text: 'hello' }] },
      }
    );
    // Span attributes are present (the event sink is exercised separately).
    assert.ok(span.attributes['gen_ai.input.messages']);
    assert.ok(span.attributes['gen_ai.output.messages']);
  });

  it('captures raw action IO only when enabled', async () => {
    const inst = new GenAiInstrumentation({ tracer, captureActionIO: true });
    const span = await runAndGetSpan(
      inst,
      info({
        metadata: {
          name: 'googleai/gemini-flash-latest',
          input: { messages: [{ role: 'user', content: [{ text: 'hi' }] }] },
        },
        labels: { 'genkit:metadata:subtype': 'model' },
      }),
      { finishReason: 'stop' }
    );
    assert.ok(span.attributes['genkit.input']);
    assert.ok(span.attributes['genkit.output']);
  });

  it('records error type and status on failure', async () => {
    const inst = new GenAiInstrumentation({ tracer });
    await assert.rejects(
      inst.runInNewSpan(
        info({
          metadata: { name: 'googleai/gemini-flash-latest' },
          labels: { 'genkit:metadata:subtype': 'model' },
        }),
        async () => {
          throw new TypeError('boom');
        }
      )
    );
    const span = exporter.getFinishedSpans().at(-1)!;
    assert.strictEqual(span.attributes['error.type'], 'TypeError');
    assert.strictEqual(span.status.code, 2 /* ERROR */);
  });

  it('warns once when no SDK is recording', async () => {
    // No injected tracer and no registered SDK -> the API returns a
    // non-recording span, so telemetry would be silently dropped.
    const inst = new GenAiInstrumentation();
    const warnings: unknown[] = [];
    const originalWarn = console.warn;
    console.warn = (...args: unknown[]) => {
      warnings.push(args[0]);
    };
    try {
      const modelInfo = info({
        metadata: { name: 'googleai/gemini-flash-latest' },
        labels: { 'genkit:metadata:subtype': 'model' },
      });
      await inst.runInNewSpan(modelInfo, async () => ({
        finishReason: 'stop',
      }));
      // A second call must not warn again.
      await inst.runInNewSpan(modelInfo, async () => ({
        finishReason: 'stop',
      }));
    } finally {
      console.warn = originalWarn;
    }
    const notRecording = warnings.filter(
      (w) => typeof w === 'string' && w.includes('no OpenTelemetry SDK')
    );
    assert.strictEqual(notRecording.length, 1);
  });

  it('exposes trace/span ids and setMetadata through the context', async () => {
    const inst = new GenAiInstrumentation({ tracer });
    let captured: GenkitSpanContext | undefined;
    await inst.runInNewSpan(
      info({
        metadata: { name: 'myFlow' },
        labels: { 'genkit:metadata:subtype': 'flow' },
      }),
      async (_span, ctx) => {
        captured = ctx;
        ctx.setMetadata({ custom: 'value' });
        return 'ok';
      }
    );
    assert.match(captured!.traceId, /^[0-9a-f]{32}$/);
    assert.match(captured!.spanId, /^[0-9a-f]{16}$/);
    const span = exporter.getFinishedSpans().at(-1)!;
    assert.strictEqual(span.attributes['genkit:metadata:custom'], 'value');
  });
});

File: py/packages/genkit-otel/tests/genai_instrumentation_test.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from collections.abc import Mapping

import pytest
from genkit_otel import ContentCapturingMode, GenAiInstrumentation
from genkit_otel._gen_ai_attributes import (
    CAPTURE_CONTENT_ENV_VAR,
    GEN_AI_OPERATION_DETAILS_EVENT,
    GenAiAttr,
    GenkitAttr,
)
from opentelemetry.trace import StatusCode

from genkit import FinishReason
from genkit._core._instrumentation_api import SpanNext
from genkit._core._model import OutputConfig
from genkit._core._typing import (
    Candidate,
    GenerationUsage,
    Part,
    Role,
    TextPart,
    ToolRequest,
    ToolRequestPart,
)
from genkit.model import Message, ModelRequest, ModelResponse
from genkit.telemetry import SpanMetadata


def _model_request(
    *,
    config: Mapping[str, object] | None = None,
    messages: list[Message] | None = None,
    output: OutputConfig | None = None,
) -> ModelRequest:
    kwargs: dict[str, object] = {
        'messages': messages or [Message(role=Role.USER, content=[Part(root=TextPart(text='hi'))])],
    }
    if config is not None:
        kwargs['config'] = config
    if output is not None:
        kwargs['output'] = output
    return ModelRequest(**kwargs)


def _model_response(
    *,
    finish_reason: FinishReason | None = None,
    message: Message | None = None,
    usage: GenerationUsage | None = None,
) -> ModelResponse:
    return ModelResponse(
        finish_reason=finish_reason or FinishReason.STOP,
        message=message or Message(role=Role.MODEL, content=[Part(root=TextPart(text='ok'))]),
        usage=usage,
    )


async def _run_model(
    instr: GenAiInstrumentation,
    name: str,
    input: object,
    fn: SpanNext[object],
) -> object:
    return await instr.run_in_new_span(
        SpanMetadata(name=name, action_type='model', input=input),
        fn,
    )


def _event_names(harness) -> list[str | None]:
    names: list[str | None] = []
    for record in harness.logs.get_finished_logs():
        inner = getattr(record, 'log_record', record)
        names.append(getattr(inner, 'event_name', None))
    return names


@pytest.mark.asyncio
async def test_emits_chat_span_with_request_provider_model(harness) -> None:
    instr = harness.instrumentation()
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(config={'temperature': 0.5, 'topK': 40, 'maxOutputTokens': 100}),
        lambda span=None: _awaitable(_model_response()),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert harness.attr(span, GenAiAttr.OPERATION_NAME) == 'chat'
    assert harness.attr(span, GenAiAttr.PROVIDER_NAME) == 'gcp.gemini'
    assert harness.attr(span, GenAiAttr.REQUEST_MODEL) == 'gemini-flash-latest'
    assert harness.attr(span, GenAiAttr.REQUEST_TEMPERATURE) == 0.5
    assert harness.attr(span, GenAiAttr.REQUEST_TOP_K) == 40
    assert harness.attr(span, GenAiAttr.REQUEST_MAX_TOKENS) == 100


@pytest.mark.asyncio
async def test_records_usage_and_finish_reason(harness) -> None:
    instr = harness.instrumentation()
    await _run_model(
        instr,
        'openai/gpt-x',
        _model_request(),
        lambda span=None: _awaitable(
            _model_response(
                finish_reason=FinishReason.LENGTH,
                usage=GenerationUsage(input_tokens=10, output_tokens=20),
            )
        ),
    )
    span = harness.span_named('chat gpt-x')
    assert span is not None
    assert harness.attr(span, GenAiAttr.PROVIDER_NAME) == 'openai'
    assert harness.attr(span, GenAiAttr.USAGE_INPUT_TOKENS) == 10
    assert harness.attr(span, GenAiAttr.USAGE_OUTPUT_TOKENS) == 20
    assert list(harness.attr(span, GenAiAttr.RESPONSE_FINISH_REASONS)) == ['length']


@pytest.mark.asyncio
async def test_reports_tool_calls_when_response_has_tool_request(harness) -> None:
    instr = harness.instrumentation()
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(),
        lambda span=None: _awaitable(
            _model_response(
                message=Message(
                    role=Role.MODEL,
                    content=[Part(root=ToolRequestPart(tool_request=ToolRequest(name='lookup', input={})))],
                )
            )
        ),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert list(harness.attr(span, GenAiAttr.RESPONSE_FINISH_REASONS)) == ['tool_calls']


@pytest.mark.asyncio
async def test_records_error_status_and_type_on_throw(harness) -> None:
    instr = harness.instrumentation()

    async def boom(span=None):
        raise RuntimeError('boom')

    with pytest.raises(RuntimeError, match='boom'):
        await _run_model(instr, 'googleai/gemini-flash-latest', _model_request(), boom)

    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert span.status.status_code == StatusCode.ERROR
    assert harness.attr(span, GenAiAttr.ERROR_TYPE) == 'RuntimeError'


@pytest.mark.asyncio
async def test_does_not_capture_content_by_default(harness) -> None:
    instr = harness.instrumentation()
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(
            messages=[Message(role=Role.USER, content=[Part(root=TextPart(text='secret'))])],
        ),
        lambda span=None: _awaitable(_model_response()),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert harness.attr(span, GenAiAttr.INPUT_MESSAGES) is None


@pytest.mark.asyncio
async def test_span_only_captures_content_on_span_not_event(harness) -> None:
    instr = harness.instrumentation(content_capturing_mode=ContentCapturingMode.SPAN_ONLY)
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(
            messages=[
                Message(role=Role.SYSTEM, content=[Part(root=TextPart(text='Be nice'))]),
                Message(role=Role.USER, content=[Part(root=TextPart(text='hello'))]),
            ]
        ),
        lambda span=None: _awaitable(_model_response()),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    input_messages = harness.attr(span, GenAiAttr.INPUT_MESSAGES)
    assert isinstance(input_messages, str)
    assert 'hello' in input_messages
    assert 'Be nice' in harness.attr(span, GenAiAttr.SYSTEM_INSTRUCTIONS)
    assert 'ok' in harness.attr(span, GenAiAttr.OUTPUT_MESSAGES)
    assert GEN_AI_OPERATION_DETAILS_EVENT not in _event_names(harness)


@pytest.mark.asyncio
async def test_event_only_emits_event_not_span_content(harness) -> None:
    instr = harness.instrumentation(content_capturing_mode=ContentCapturingMode.EVENT_ONLY)
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(messages=[Message(role=Role.USER, content=[Part(root=TextPart(text='hello'))])]),
        lambda span=None: _awaitable(_model_response()),
    )
    assert GEN_AI_OPERATION_DETAILS_EVENT in _event_names(harness)
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert harness.attr(span, GenAiAttr.INPUT_MESSAGES) is None


@pytest.mark.asyncio
async def test_span_and_event_writes_both(harness) -> None:
    instr = harness.instrumentation(content_capturing_mode=ContentCapturingMode.SPAN_AND_EVENT)
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(messages=[Message(role=Role.USER, content=[Part(root=TextPart(text='hello'))])]),
        lambda span=None: _awaitable(_model_response()),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert 'hello' in harness.attr(span, GenAiAttr.INPUT_MESSAGES)
    assert GEN_AI_OPERATION_DETAILS_EVENT in _event_names(harness)


@pytest.mark.asyncio
async def test_nests_flow_and_model_into_one_trace(harness) -> None:
    instr = harness.instrumentation()

    async def flow_body(span=None):
        return await _run_model(
            instr,
            'googleai/gemini-flash-latest',
            _model_request(),
            lambda inner=None: _awaitable(_model_response()),
        )

    await instr.run_in_new_span(SpanMetadata(name='myFlow', action_type='flow'), flow_body)
    flow = harness.span_named('myFlow')
    model = harness.span_named('chat gemini-flash-latest')
    assert flow is not None and model is not None
    assert harness.attr(flow, GenkitAttr.ACTION_TYPE) == 'flow'
    assert model.context.trace_id == flow.context.trace_id


@pytest.mark.asyncio
async def test_execute_tool_spans_only_when_enabled(harness) -> None:
    off = harness.instrumentation()

    async def sunny(span=None):
        return 'sunny'

    await off.run_in_new_span(SpanMetadata(name='weather', action_type='tool'), sunny)
    assert harness.span_named('execute_tool weather') is None

    on = harness.instrumentation(emit_tool_spans=True)
    await on.run_in_new_span(SpanMetadata(name='weather', action_type='tool'), sunny)
    span = harness.span_named('execute_tool weather')
    assert span is not None
    assert harness.attr(span, GenAiAttr.OPERATION_NAME) == 'execute_tool'
    assert harness.attr(span, GenAiAttr.TOOL_NAME) == 'weather'


@pytest.mark.asyncio
async def test_does_not_capture_action_io_by_default(harness) -> None:
    instr = harness.instrumentation()
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(),
        lambda span=None: _awaitable(_model_response()),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert harness.attr(span, GenkitAttr.INPUT) is None
    assert harness.attr(span, GenkitAttr.OUTPUT) is None


@pytest.mark.asyncio
async def test_capture_action_io_records_raw_io_on_all_span_types(harness) -> None:
    instr = harness.instrumentation(capture_action_io=True, emit_tool_spans=True)

    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(),
        lambda span=None: _awaitable(_model_response()),
    )
    model = harness.span_named('chat gemini-flash-latest')
    assert model is not None
    assert isinstance(harness.attr(model, GenkitAttr.INPUT), str)
    assert isinstance(harness.attr(model, GenkitAttr.OUTPUT), str)

    async def sunny(span=None):
        return 'sunny'

    await instr.run_in_new_span(
        SpanMetadata(name='weather', action_type='tool', input='Paris'),
        sunny,
    )
    tool = harness.span_named('execute_tool weather')
    assert tool is not None
    assert 'Paris' in harness.attr(tool, GenkitAttr.INPUT)
    assert 'sunny' in harness.attr(tool, GenkitAttr.OUTPUT)
    assert harness.attr(tool, GenAiAttr.INPUT_MESSAGES) is None
    assert harness.attr(tool, GenAiAttr.OUTPUT_MESSAGES) is None

    async def flow_out(span=None):
        return 'out'

    await instr.run_in_new_span(SpanMetadata(name='myFlow', action_type='flow', input='in'), flow_out)
    flow = harness.span_named('myFlow')
    assert flow is not None
    assert 'in' in harness.attr(flow, GenkitAttr.INPUT)
    assert 'out' in harness.attr(flow, GenkitAttr.OUTPUT)
    assert harness.attr(flow, GenAiAttr.INPUT_MESSAGES) is None


@pytest.mark.asyncio
async def test_captures_legacy_candidates_message(harness) -> None:
    instr = harness.instrumentation(content_capturing_mode=ContentCapturingMode.SPAN_ONLY)
    legacy = ModelResponse(
        finish_reason=FinishReason.STOP,
        candidates=[
            Candidate(
                index=0,
                finish_reason=FinishReason.STOP,
                message=Message(role=Role.MODEL, content=[Part(root=TextPart(text='legacy answer'))]),
            )
        ],
    )
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(),
        lambda span=None: _awaitable(legacy),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert 'legacy answer' in harness.attr(span, GenAiAttr.OUTPUT_MESSAGES)


@pytest.mark.asyncio
async def test_reports_tool_calls_from_legacy_candidates(harness) -> None:
    instr = harness.instrumentation()
    legacy = ModelResponse(
        finish_reason=FinishReason.STOP,
        candidates=[
            Candidate(
                index=0,
                finish_reason=FinishReason.STOP,
                message=Message(
                    role=Role.MODEL,
                    content=[Part(root=ToolRequestPart(tool_request=ToolRequest(name='lookup', input={})))],
                ),
            )
        ],
    )
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(),
        lambda span=None: _awaitable(legacy),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert list(harness.attr(span, GenAiAttr.RESPONSE_FINISH_REASONS)) == ['tool_calls']


@pytest.mark.asyncio
async def test_classifies_action_subtype_model_as_chat(harness) -> None:
    """Python action leftover: type=action, subtype=model still becomes a chat span."""
    instr = harness.instrumentation()

    async def body(span=None):
        return _model_response()

    await instr.run_in_new_span(
        SpanMetadata(
            name='googleai/gemini-flash-latest',
            action_type='action',
            subtype='model',
            input=_model_request(),
        ),
        body,
    )
    assert harness.span_named('chat gemini-flash-latest') is not None
    assert harness.span_named('googleai/gemini-flash-latest') is None


@pytest.mark.asyncio
async def test_explicit_mode_overrides_env(harness, monkeypatch: pytest.MonkeyPatch) -> None:
    monkeypatch.setenv(CAPTURE_CONTENT_ENV_VAR, 'SPAN_ONLY')
    instr = harness.instrumentation(content_capturing_mode=ContentCapturingMode.NO_CONTENT)
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(messages=[Message(role=Role.USER, content=[Part(root=TextPart(text='secret'))])]),
        lambda span=None: _awaitable(_model_response()),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert harness.attr(span, GenAiAttr.INPUT_MESSAGES) is None


@pytest.mark.asyncio
async def test_invalid_env_token_defaults_to_no_content(harness, monkeypatch: pytest.MonkeyPatch) -> None:
    monkeypatch.setenv(CAPTURE_CONTENT_ENV_VAR, 'bogus')
    instr = GenAiInstrumentation(
        tracer=harness.tracer_provider.get_tracer('test'),
        meter=harness.meter_provider.get_meter('test'),
        otel_logger=harness.logger_provider.get_logger('test'),
    )
    assert instr.content_capturing_mode is ContentCapturingMode.NO_CONTENT


@pytest.mark.asyncio
async def test_output_type_json_from_request(harness) -> None:
    instr = harness.instrumentation()
    await _run_model(
        instr,
        'googleai/gemini-flash-latest',
        _model_request(output=OutputConfig(format='json')),
        lambda span=None: _awaitable(_model_response()),
    )
    span = harness.span_named('chat gemini-flash-latest')
    assert span is not None
    assert harness.attr(span, GenAiAttr.OUTPUT_TYPE) == 'json'


async def _awaitable(value: object) -> object:
    return value
